src/Storages/MergeTree/MergeTreePartsMover.cpp

```diff
@@ -210,71 +212,93 @@ bool MergeTreePartsMover::selectPartsForMove(
+
+    const bool is_zero_copy = disk->supportZeroCopyReplication() && settings->allow_remote_fs_zero_copy_replication;
+
+    const fs::path moving_dir = fs::path(data->getRelativeDataPath()) / MergeTreeData::MOVING_DIR_NAME;
+    const String part_dir = storage.getPartDirectory();
+    const fs::path result_dir = moving_dir / part_dir;
+
+    if (is_vfs || is_zero_copy)
     {
-        /// Try zero-copy replication and fallback to default copy if it's not possible
-        moving_part.part->assertOnDisk();
-        String path_to_clone = fs::path(data->getRelativeDataPath()) / MergeTreeData::MOVING_DIR_NAME / "";
-        String relative_path = part->getDataPartStorage().getPartDirectory();
-        if (disk->exists(path_to_clone + relative_path))
+        part->assertOnDisk();
+        if (disk->exists(result_dir))
         {
+            const String full_path = fullPath(disk, result_dir);
             // If setting is on, we should've already cleaned moving/ dir on startup
             if (data->allowRemoveStaleMovingParts())
                 throw Exception(ErrorCodes::DIRECTORY_ALREADY_EXISTS,
                     "Cannot clone part {} from '{}' to '{}': path '{}' already exists",
-                    part->name, part->getDataPartStorage().getDiskName(), disk->getName(),
-                    fullPath(disk, path_to_clone + relative_path));
+                    part->name, storage.getDiskName(), disk->getName(),
+                    full_path);
 
-            LOG_DEBUG(log, "Path {} already exists. Will remove it and clone again",
-                fullPath(disk, path_to_clone + relative_path));
-            disk->removeRecursive(fs::path(path_to_clone) / relative_path / "");
+            LOG_DEBUG(log, "Path {} already exists. Will remove it and clone again", full_path);
+            disk->removeRecursive(result_dir);
         }
 
-        disk->createDirectories(path_to_clone);
+        disk->createDirectories(moving_dir);
+    }
 
-        auto zero_copy_part = data->tryToFetchIfShared(*part, disk, fs::path(path_to_clone) / part->name);
+    MutableDataPartStoragePtr cloned_part_storage;
 
-        if (zero_copy_part)
-        {
-            /// FIXME for some reason we cannot just use this part, we have to re-create it through MergeTreeDataPartBuilder
-            zero_copy_part->is_temp = false;    /// Do not remove it in dtor
-            cloned_part_storage = zero_copy_part->getDataPartStoragePtr();
-        }
-        else
-        {
-            LOG_INFO(log, "Part {} was not fetched, we are the first who move it to another disk, so we will copy it", part->name);
-            cloned_part_storage = part->getDataPartStorage().clonePart(
-                path_to_clone, part->getDataPartStorage().getPartDirectory(), disk, read_settings, write_settings, log, cancellation_hook);
-        }
+    if (is_vfs &&
+        (vfs_downloaded_metadata = vfs_disk->tryDownloadMetadata(vfs_move_metadata_key, result_dir)))
```

Comment:
It looks like we does not add new link in vfs log after downloading of already moved parts. This breaks gc loop here
https://github.com/ClickHouse/ClickHouse/blob/e601ad10e7f7898a7520c66ef6e63747ec4685f9/src/Disks/ObjectStorages/VFSGarbageCollector.cpp#L207-L212
This might be reason of failing tests involving moves.

Reply:
Fixed